### PR TITLE
Optimise streaming

### PR DESCRIPTION
**Pull request by ngsp94**
Streaming has been optimised with the following changes:
- Some calculations have been changed to variables to reduce calculation and function calls: frame height and width, frame buffer size
- Split screen implementation has changed; calling a subset of pixels from viewport instead of splitting pixels manually
- Declaration of 'new' variable brought out of loop
- Bit shifting is used instead of multiplication in converting pixels to bytes

### CODE REVIEW

File: Source/CloudyPanelPlugin/Private/CloudyPanelPlugin.cpp [hunk 1]
```diff
@@ -266,77 +257,45 @@ void CCloudyPanelPluginModule::StreamFrameToClient() {
...
 
-}
+	FViewport* ReadingViewport = GEngine->GameViewport->Viewport;
 
+	FIntRect Screen1 = FIntRect(0, 0, halfSizeX, halfSizeY);
+	FIntRect Screen2 = FIntRect(halfSizeX, 0, sizeX, halfSizeY);
+	FIntRect Screen3 = FIntRect(0, halfSizeY, halfSizeX, sizeY);
+	FIntRect Screen4 = FIntRect(halfSizeX, halfSizeY, sizeX, sizeY);
+	FReadSurfaceDataFlags flags = FReadSurfaceDataFlags(ERangeCompressionMode::RCM_MinMaxNorm,
+		ECubeFace::CubeFace_NegX);
```
Comment (joelngwt) [inline Source/CloudyPanelPlugin/Private/CloudyPanelPlugin.cpp]: This function is being called once per frame. Can this initialization be shifted out of the function so it does not need to be done every frame?

File: Source/CloudyPanelPlugin/Private/CloudyPanelPlugin.cpp [hunk 2]
```diff
@@ -266,77 +257,45 @@ void CCloudyPanelPluginModule::StreamFrameToClient() {
...
 	FColor Pixel;
-	for (int i = 0; i < NumberOfPlayers; i++) {
-		PixelBuffer = new uint32[sizeX * sizeY * PIXEL_SIZE];
-		//UE_LOG(ModuleLog, Warning, TEXT("FrameIndex %d"), FrameIndex);
-		for (int j = 0; j < FrameBufferList[i].Num(); j++) {
+	PixelBuffer = new uint32[sizeX * sizeY * PIXEL_SIZE];
+
+	for (int i = 0; i < NumberOfPlayers; i++) {	
+		int FrameSize = FrameBufferList[i].Num();
+		for (int j = 0; j < FrameSize; j++) {
```
Comment (joelngwt) [inline Source/CloudyPanelPlugin/Private/CloudyPanelPlugin.cpp]: Try using ++j instead. The difference might be significant due to the FrameSize.

**Comment by ngsp94** [comment]
More changes:
- Can capture stream from connected devices using own IP
- Changes according to comments earlier, pulling out unnecessary statements, ++j
- Separated per-player actions and setup actions (only need to call once for set up)
- Tuned the ffmpeg parameters for speed & quality. Can push 30 fps from Unreal, but receiving side only around 10-15 fps

**Comment by joelngwt** [comment]
@ngsp94 Do you want this to be merged? It is really old

**Comment by ngsp94** [comment]
Well, this whole part has been taken out in the refactor branch, so actually it doesn't matter after the newer branch is merged

**Comment by joelngwt** [comment]
Closing this. Will be merged with the newer branch,

**Pull request closed without merging**